statsd/statsd.go

```diff
@@ -395,44 +382,12 @@ func (c *Client) namespace() string {
 	return ""
 }
 
-func (c *Client) writeMetricUnsafe(m metric) error {
-	switch m.metricType {
-	case gauge:
-		return c.buffer.writeGauge(m.namespace, m.globalTags, m.name, m.fvalue, m.tags, m.rate)
-	case count:
-		return c.buffer.writeCount(m.namespace, m.globalTags, m.name, m.ivalue, m.tags, m.rate)
-	case histogram:
-		return c.buffer.writeHistogram(m.namespace, m.globalTags, m.name, m.fvalue, m.tags, m.rate)
-	case distribution:
-		return c.buffer.writeDistribution(m.namespace, m.globalTags, m.name, m.fvalue, m.tags, m.rate)
-	case set:
-		return c.buffer.writeSet(m.namespace, m.globalTags, m.name, m.svalue, m.tags, m.rate)
-	case timing:
-		return c.buffer.writeTiming(m.namespace, m.globalTags, m.name, m.fvalue, m.tags, m.rate)
-	case event:
-		return c.buffer.writeEvent(*m.evalue, m.globalTags)
-	case serviceCheck:
-		return c.buffer.writeServiceCheck(*m.scvalue, m.globalTags)
-	default:
-		return nil
-	}
-}
-
 func (c *Client) addMetric(m metric) error {
 	if c == nil {
 		return ErrNoClient
 	}
-	if c.shouldSample(m.rate) {
-		return nil
-	}
-	c.Lock()
-	var err error
-	if err = c.writeMetricUnsafe(m); err == errBufferFull {
-		c.flushUnsafe()
-		err = c.writeMetricUnsafe(m)
-	}
-	c.Unlock()
-	return err
+	h := hashString32(m.name)
+	return c.bufferShards[h%uint32(len(c.bufferShards))].processMetric(m)
```

Comment:
The `fastrange` is fair only if we have a wide spread of value. We might have similar value in our case since many metrics will share the majority of there name.

For example, running some very simple test with our benchmark (we have GOMAXPROC different metric name, so in my case `test.metric1` to `test.metric8`):

This produce this spread with `%`:
```
modulo test.metric1 -> shard 28
modulo test.metric2 -> shard 21
modulo test.metric3 -> shard 2
modulo test.metric4 -> shard 3
modulo test.metric5 -> shard 16
modulo test.metric6 -> shard 9
modulo test.metric7 -> shard 22
modulo test.metric8 -> shard 7
```

This produce this spread with `fastrange`:
```
fastrange test.metric1 -> shard 14
fastrange test.metric2 -> shard 14
fastrange test.metric3 -> shard 14
fastrange test.metric4 -> shard 14
fastrange test.metric5 -> shard 14
fastrange test.metric6 -> shard 14
fastrange test.metric7 -> shard 14
fastrange test.metric8 -> shard 15
```

Having fully randomized metrics name give the same spread than the modulo but does not speed things up much:

fully randomize metric name + modulo:
```
BenchmarkStatsdUDP-8   	22229242	        52.7 ns/op
BenchmarkStatsdUDS-8   	21336699	        55.6 ns/op
```

fully randomize metric name + fast range:
```
BenchmarkStatsdUDP-8   	23752176	        51.0 ns/op
BenchmarkStatsdUDS-8   	24448677	        50.9 ns/op
```

Reply:
modulo it is! But thanks for doing this. It's interesting how we could not reproduce at all the kind of gains `fastrange` claims. Just curious how many buckets did we have here?

Reply:
The default it 32 (https://github.com/DataDog/datadog-go/pull/108/files#diff-4471ec43bb44996af266fa51c1a2acf3R22)